### PR TITLE
Jacobian Op

### PR DESCRIPTION
**Pull request by jd-bartlett96**
Jacobian(y, x) currently only supports inputs for x of rank <= 1. Looking into supporting higher dimensional wrt inputs but this op should still be useful for the majority of use cases I can think of (which would mostly be 1d inputs for both y and x).

Run tests with `bazelisk run //networks/scitile/pd_solve:gradtest`

### CODE REVIEW

File: tile/lang/ast/jacobian.cc [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+  std::stack<ExprPtr> stack_;
+  std::unordered_set<const Expr*> seen_;
+  std::unordered_map<const Expr*, std::vector<UseInfo>> uses_;
+};
+
+class Jacobian {
+ public:
+  explicit Jacobian(const ExprPtr& err) : uses_(err) {
+    IVLOG(2, "Gradient::Gradient> err: " << err);
+    // LogicalShape* into = LogicalShape(DataType::FLOAT32, {3});m
```
Comment (tzerrell) [inline tile/lang/ast/jacobian.cc:88]: Can this be removed?

**Review by tzerrell — COMMENTED** [review]
Cool!

File: networks/scitile/pd_solve/gradtest.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+import unittest
+import numpy.testing as npt
+
+
+def matmul_2_1(A, B):
+    I, J = edsl.TensorDims(2)
+    i, j = edsl.TensorIndexes(2)
+    A.bind_dims(I, J)
+    B.bind_dims(J)
```
[the remark this answers is not in the document]
Comment (flaub) [reply networks/scitile/pd_solve/gradtest.py:17]: The 2nd bind is for asserting that the size of that dimension is the same as it was previously defined in another tensor.

File: tile/lang/ast/jacobian.cc (hunk 1 above)
Comment (jd-bartlett96) [reply tile/lang/ast/jacobian.cc]: Yep! Sorry, missed this and a few other outdated lines/comments in my cleanup

File: networks/scitile/pd_solve/gradtest.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tzerrell) [reply networks/scitile/pd_solve/gradtest.py:17]: oh, great, that does exactly what I would hope for then. Thanks!

**Comment by jd-bartlett96** [comment]
Changing to WIP until I can fix higher-rank tensor functionality & throw errors for any cases not supported (I've found a number of cases that get passes through the op without throwing any error/warning then return an incorrect result).

**Comment by jd-bartlett96** [comment]
Op now should work properly for the majority of useful cases. The few remaining use cases that are not implemented should throw a error message letting the user know that functionality is not yet implemented.

**Review by tzerrell — APPROVED** [review]
Nice, LGTM

**Review by tzerrell — APPROVED** [review]

**Pull request merged**